File: internal/queue/queue.go
package queue

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"sort"
	"time"
)

// PR represents a pull request in the merge queue.
type PR struct {
	Number    int
	HeadRef   string
	HeadSHA   string
	Title     string
	CreatedAt int64 // unix timestamp for ordering
}

// LabelState represents the lifecycle of a PR in the queue.
type LabelState string

const (
	StatePending LabelState = ""       // just the queue label
	StateActive  LabelState = "active" // being processed
	StateFailed  LabelState = "failed" // CI failed or conflict
)

// QueueLabel returns the full label string for a given state.
func QueueLabel(base string, state LabelState) string {
	if state == StatePending {
		return base
	}
	return fmt.Sprintf("%s:%s", base, string(state))
}

// GitHubAPI defines the interface for GitHub operations needed by the queue.
type GitHubAPI interface {
	// ListPRsWithLabel returns open PRs that have the given label, oldest first.
	ListPRsWithLabel(ctx context.Context, label string) ([]PR, error)

	// AddLabel adds a label to a PR.
	AddLabel(ctx context.Context, prNumber int, label string) error

	// RemoveLabel removes a label from a PR.
	RemoveLabel(ctx context.Context, prNumber int, label string) error

	// Comment posts a comment on a PR.
	Comment(ctx context.Context, prNumber int, body string) error

	// CreateLabel creates a label in the repository.
	CreateLabel(ctx context.Context, name string, color string, description string) error
}

// WorkflowAPI defines the interface for workflow dispatch and polling.
// Kept separate from GitHubAPI to reduce coupling — the queue package
// does not use these methods.
type WorkflowAPI interface {
	// TriggerWorkflow dispatches a workflow.
	TriggerWorkflow(ctx context.Context, workflowFile string, ref string, inputs map[string]interface{}) error

	// GetWorkflowRunStatus polls for the latest run of a workflow on a ref
	// created after dispatchedAt and returns its conclusion.
	GetWorkflowRunStatus(ctx context.Context, workflowFile string, ref string, dispatchedAt time.Time) (conclusion string, err error)

	// ClosePR closes a pull request.
	ClosePR(ctx context.Context, prNumber int) error
}

// Queue manages the merge queue state machine.
type Queue struct {
	api      GitHubAPI
	label    string
	dryRun   bool
	logFunc  func(string, ...any)
}

// New creates a new Queue.
func New(api GitHubAPI, label string, dryRun bool, logFunc func(string, ...any)) *Queue {
	return &Queue{
		api:     api,
		label:   label,
		dryRun:  dryRun,
		logFunc: logFunc,
	}
}

// Collect returns open PRs with the queue label, sorted oldest first.
func (q *Queue) Collect(ctx context.Context) ([]PR, error) {
	prs, err := q.api.ListPRsWithLabel(ctx, q.label)
	if err != nil {
		return nil, fmt.Errorf("listing queued PRs: %w", err)
	}
	sort.Slice(prs, func(i, j int) bool {
		return prs[i].CreatedAt < prs[j].CreatedAt
	})
	return prs, nil
}

// Activate transitions PRs from pending to active state.
func (q *Queue) Activate(ctx context.Context, prs []PR) error {
	for _, pr := range prs {
		q.logFunc("Activating PR #%d", pr.Number)
		if q.dryRun {
			continue
		}
		if err := q.api.AddLabel(ctx, pr.Number, QueueLabel(q.label, StateActive)); err != nil {
			return fmt.Errorf("adding active label to #%d: %w", pr.Number, err)
		}
		if err := q.api.RemoveLabel(ctx, pr.Number, QueueLabel(q.label, StatePending)); err != nil && !isNotFoundError(err) {
			return fmt.Errorf("removing pending label from #%d: %w", pr.Number, err)
		}
	}
	return nil
}

// MarkFailed transitions a PR to the failed state and posts a comment.
func (q *Queue) MarkFailed(ctx context.Context, pr PR, reason string) error {
	q.logFunc("Marking PR #%d as failed: %s", pr.Number, reason)
	if q.dryRun {
		return nil
	}
	// Remove active label if present (ignore 404)
	if err := q.api.RemoveLabel(ctx, pr.Number, QueueLabel(q.label, StateActive)); err != nil && !isNotFoundError(err) {
		return fmt.Errorf("removing active label from #%d: %w", pr.Number, err)
	}
	// Remove pending label if present (ignore 404)
	if err := q.api.RemoveLabel(ctx, pr.Number, QueueLabel(q.label, StatePending)); err != nil && !isNotFoundError(err) {
		return fmt.Errorf("removing pending label from #%d: %w", pr.Number, err)
	}
	if err := q.api.AddLabel(ctx, pr.Number, QueueLabel(q.label, StateFailed)); err != nil {
		return fmt.Errorf("adding failed label to #%d: %w", pr.Number, err)
	}
	if err := q.api.Comment(ctx, pr.Number, fmt.Sprintf("Merge queue: %s", reason)); err != nil {
		return fmt.Errorf("commenting on #%d: %w", pr.Number, err)
	}
	return nil
}

// Requeue moves a PR back to pending state.
func (q *Queue) Requeue(ctx context.Context, pr PR) error {
	q.logFunc("Requeuing PR #%d", pr.Number)
	if q.dryRun {
		return nil
	}
	if err := q.api.RemoveLabel(ctx, pr.Number, QueueLabel(q.label, StateActive)); err != nil && !isNotFoundError(err) {
		return fmt.Errorf("removing active label from #%d: %w", pr.Number, err)
	}
	if err := q.api.RemoveLabel(ctx, pr.Number, QueueLabel(q.label, StateFailed)); err != nil && !isNotFoundError(err) {
		return fmt.Errorf("removing failed label from #%d: %w", pr.Number, err)
	}
	if err := q.api.AddLabel(ctx, pr.Number, QueueLabel(q.label, StatePending)); err != nil {
		return fmt.Errorf("requeuing #%d: %w", pr.Number, err)
	}
	return nil
}

// SetupLabels creates the queue labels in the repository.
func (q *Queue) SetupLabels(ctx context.Context) error {
	labels := []struct {
		name  string
		color string
		desc  string
	}{
		{QueueLabel(q.label, StatePending), "0e8a16", "PR is queued for merging"},
		{QueueLabel(q.label, StateActive), "1d76db", "PR is being processed by merge queue"},
		{QueueLabel(q.label, StateFailed), "d93f0b", "PR failed in merge queue"},
	}
	for _, l := range labels {
		q.logFunc("Creating label %q", l.name)
		if q.dryRun {
			continue
		}
		if err := q.api.CreateLabel(ctx, l.name, l.color, l.desc); err != nil {
			// Ignore "already_exists" errors to make setup idempotent
			if !isAlreadyExistsError(err) {
				return fmt.Errorf("creating label %q: %w", l.name, err)
			}
			q.logFunc("Label %q already exists, skipping", l.name)
		}
	}
	return nil
}

// HTTPError is an interface for errors that carry an HTTP status code.
type HTTPError interface {
	error
	StatusCode() int
}

// AlreadyExistsError is an optional interface for errors that indicate
// a resource already exists (e.g. duplicate label).
type AlreadyExistsError interface {
	AlreadyExists() bool
}

func isAlreadyExistsError(err error) bool {
	var ae AlreadyExistsError
	if errors.As(err, &ae) {
		return ae.AlreadyExists()
	}
	return false
}

func isNotFoundError(err error) bool {
	var httpErr HTTPError
	if errors.As(err, &httpErr) {
		return httpErr.StatusCode() == http.StatusNotFound
	}
	return false
}

File: src/queue.ts
/** PR represents a pull request in the merge queue. */
export interface PR {
  number: number;
  headRef: string;
  headSHA: string;
  title: string;
  state?: "open" | "closed";
  createdAt: number; // unix timestamp for ordering
  /**
   * Current label names on the PR. Optional because the issues-list
   * code path filters by label and so callers don't need them; populated
   * by `getPR` so callers can re-validate label state on a single PR.
   */
  labels?: string[];
}

export type LabelState = "" | "active" | "failed";

export const STATE_PENDING: LabelState = "";
export const STATE_ACTIVE: LabelState = "active";
export const STATE_FAILED: LabelState = "failed";

/**
 * Default cap on how many times a single PR may be requeued before the queue
 * gives up and marks it failed. This is the backstop that GUARANTEES a
 * deterministic failure can never re-trigger the workflow forever: error
 * classification (ConfigurationError, HTTP 404/422, …) can never be complete
 * over an open-ended failure space, so a root-cause-agnostic attempt cap
 * bounds every requeue path — including ones whose permanent failure slips
 * past classification.
 */
export const MAX_REQUEUE_ATTEMPTS = 10;

const ATTEMPT_INFIX = ":attempt-";

/** The attempt-counter label for a given count, e.g. `queue:attempt-3`. */
export function attemptLabel(base: string, n: number): string {
  return `${base}${ATTEMPT_INFIX}${n}`;
}

/**
 * Reads the highest requeue-attempt count encoded in a PR's labels
 * (`<base>:attempt-N`), returning 0 when none is present, alongside the
 * concrete attempt-label names found so callers can clear them. Tolerates
 * multiple/stale attempt labels by taking the max.
 */
export function readAttemptCount(
  base: string,
  labels: string[] | undefined,
): { count: number; labels: string[] } {
  const prefix = `${base}${ATTEMPT_INFIX}`;
  let count = 0;
  const found: string[] = [];
  for (const l of labels ?? []) {
    if (!l.startsWith(prefix)) continue;
    found.push(l);
    const n = Number(l.slice(prefix.length));
    if (Number.isInteger(n) && n > count) count = n;
  }
  return { count, labels: found };
}

/** Returns the full label string for a given state. */
export function queueLabel(base: string, state: LabelState): string {
  if (state === "") return base;
  return `${base}:${state}`;
}

/** Identifies a workflow run that has been dispatched. */
export interface WorkflowRunHandle {
  runId: number;
  htmlUrl: string;
}

/** Final result of a workflow run once it has completed. */
export interface WorkflowRunResult {
  conclusion: string;
  htmlUrl: string;
}

/** GitHubAPI defines the interface for GitHub operations needed by the queue. */
export interface GitHubAPI {
  listPRsWithLabel(label: string, limit: number): Promise<PR[]>;
  addLabel(prNumber: number, label: string): Promise<void>;
  removeLabel(prNumber: number, label: string): Promise<void>;
  comment(prNumber: number, body: string): Promise<void>;
  createLabel(name: string, color: string, description: string): Promise<void>;
}

/** WorkflowAPI defines the interface for workflow dispatch and polling. */
export interface WorkflowAPI {
  triggerWorkflow(
    workflowFile: string,
    ref: string,
    inputs?: Record<string, string>,
  ): Promise<void>;
  /** Waits for the dispatched workflow run to appear and returns its URL. */
  findWorkflowRun(
    workflowFile: string,
    ref: string,
    dispatchedAt: Date,
    headSha?: string,
  ): Promise<WorkflowRunHandle>;
  /** Polls an already-located run until it completes. */
  waitForWorkflowRun(runId: number): Promise<WorkflowRunResult>;
  closePR(prNumber: number): Promise<void>;
}

function isNotFoundError(err: unknown): boolean {
  return (
    typeof err === "object" &&
    err !== null &&
    "status" in err &&
    (err as { status: number }).status === 404
  );
}

function isAlreadyExistsError(err: unknown): boolean {
  if (typeof err !== "object" || err === null) return false;
  const e = err as Record<string, unknown>;
  if (e.status !== 422) return false;
  const resp = e.response as Record<string, unknown> | undefined;
  const data = resp?.data as Record<string, unknown> | undefined;
  const errors = data?.errors;
  if (!Array.isArray(errors)) return false;
  return errors.some(
    (error: Record<string, unknown>) => error.code === "already_exists",
  );
}

type LogFunc = (msg: string) => void;

/** Queue manages the merge queue label state machine. Comment composition lives at the orchestration layer. */
export class Queue {
  private api: GitHubAPI;
  private label: string;
  private dryRun: boolean;
  private log: LogFunc;
  private maxAttempts: number;

  constructor(
    api: GitHubAPI,
    label: string,
    dryRun: boolean,
    log?: LogFunc,
    maxAttempts: number = MAX_REQUEUE_ATTEMPTS,
  ) {
    this.api = api;
    this.label = label;
    this.dryRun = dryRun;
    this.log = log ?? (() => {});
    // Guard against a misconfigured 0/negative cap silently disabling the
    // backstop; fall back to the default so the loop stays bounded.
    this.maxAttempts =
      Number.isInteger(maxAttempts) && maxAttempts > 0
        ? maxAttempts
        : MAX_REQUEUE_ATTEMPTS;
  }

  /** The effective requeue cap (after validation), for comment text. */
  get attemptCap(): number {
    return this.maxAttempts;
  }

  /** Removes every `<base>:attempt-N` label currently known on the PR. */
  private async clearAttemptLabels(pr: PR): Promise<void> {
    const { labels } = readAttemptCount(this.label, pr.labels);
    for (const l of labels) {
      try {
        await this.api.removeLabel(pr.number, l);
      } catch (err) {
        if (!isNotFoundError(err)) throw err;
      }
    }
  }

  /** Returns open PRs with the queue label, sorted oldest first. */
  async collect(limit: number): Promise<PR[]> {
    const prs = await this.api.listPRsWithLabel(this.label, limit);
    prs.sort((a, b) => a.createdAt - b.createdAt);
    return prs;
  }

  /** Transitions PRs from pending to active state. */
  async activate(prs: PR[]): Promise<void> {
    for (const pr of prs) {
      this.log(`Activating PR #${pr.number}`);
      if (this.dryRun) continue;
      await this.api.addLabel(
        pr.number,
        queueLabel(this.label, STATE_ACTIVE),
      );
      try {
        await this.api.removeLabel(
          pr.number,
          queueLabel(this.label, STATE_PENDING),
        );
      } catch (err) {
        if (!isNotFoundError(err)) throw err;
      }
      // Clear any lingering queue:failed label — a PR re-entering the queue
      // after a failure has the base label re-added by the author, but
      // activate() must clean up the old failed state so it doesn't persist
      // after a subsequent successful merge.
      try {
        await this.api.removeLabel(
          pr.number,
          queueLabel(this.label, STATE_FAILED),
        );
      } catch (err) {
        if (!isNotFoundError(err)) throw err;
      }
    }
  }

  /** Transitions a PR to the failed state. */
  async markFailed(pr: PR, reason: string): Promise<void> {
    this.log(`Marking PR #${pr.number} as failed: ${reason}`);
    if (this.dryRun) return;
    try {
      await this.api.removeLabel(
        pr.number,
        queueLabel(this.label, STATE_ACTIVE),
      );
    } catch (err) {
      if (!isNotFoundError(err)) throw err;
    }
    try {
      await this.api.removeLabel(
        pr.number,
        queueLabel(this.label, STATE_PENDING),
      );
    } catch (err) {
      if (!isNotFoundError(err)) throw err;
    }
    // Reset the requeue-attempt counter: a PR leaving the queue (failed or,
    // later, re-added by the author) must start from a fresh budget.
    await this.clearAttemptLabels(pr);
    await this.api.addLabel(
      pr.number,
      queueLabel(this.label, STATE_FAILED),
    );
  }

  /**
   * Moves a PR back to pending state so the queue retries it — UNLESS it has
   * already been requeued `maxAttempts` times, in which case it is marked
   * failed instead. Returns `true` if the PR was requeued, `false` if the
   * attempt cap was reached (the PR is now in the failed state).
   *
   * The cap is the root-cause-agnostic backstop: even a permanent failure
   * that error classification fails to recognise can re-enter the queue at
   * most `maxAttempts` times before this turns it into the terminal
   * `failed` state, which the workflow's label trigger no longer matches.
   */
  async requeue(pr: PR): Promise<boolean> {
    const { count } = readAttemptCount(this.label, pr.labels);
    const next = count + 1;
    if (next > this.maxAttempts) {
      this.log(
        `PR #${pr.number} reached the requeue cap (${this.maxAttempts}); marking failed instead of requeueing`,
      );
      await this.markFailed(
        pr,
        `exceeded ${this.maxAttempts} requeue attempts`,
      );
      return false;
    }
    this.log(
      `Requeuing PR #${pr.number} (attempt ${next}/${this.maxAttempts})`,
    );
    if (this.dryRun) return true;
    // Bump the attempt counter: drop any prior attempt labels, add the new
    // one. Done before re-adding the base label so the count is durable the
    // moment the PR re-enters the queue.
    await this.clearAttemptLabels(pr);
    await this.api.addLabel(pr.number, attemptLabel(this.label, next));
    try {
      await this.api.removeLabel(
        pr.number,
        queueLabel(this.label, STATE_ACTIVE),
      );
    } catch (err) {
      if (!isNotFoundError(err)) throw err;
    }
    try {
      await this.api.removeLabel(
        pr.number,
        queueLabel(this.label, STATE_FAILED),
      );
    } catch (err) {
      if (!isNotFoundError(err)) throw err;
    }
    await this.api.addLabel(
      pr.number,
      queueLabel(this.label, STATE_PENDING),
    );
    return true;
  }

  /** Creates the queue labels in the repository. */
  async setupLabels(): Promise<void> {
    const labels = [
      {
        name: queueLabel(this.label, STATE_PENDING),
        color: "0e8a16",
        desc: "PR is queued for merging",
      },
      {
        name: queueLabel(this.label, STATE_ACTIVE),
        color: "1d76db",
        desc: "PR is being processed by merge queue",
      },
      {
        name: queueLabel(this.label, STATE_FAILED),
        color: "d93f0b",
        desc: "PR failed in merge queue",
      },
    ];

    for (const l of labels) {
      this.log(`Creating label "${l.name}"`);
      if (this.dryRun) continue;
      try {
        await this.api.createLabel(l.name, l.color, l.desc);
      } catch (err) {
        if (!isAlreadyExistsError(err)) throw err;
        this.log(`Label "${l.name}" already exists, skipping`);
      }
    }
  }
}
